ops/lib/tls_certificates/ca_client.py

```diff
@@ -0,0 +1,161 @@
+            # certificates to be generated by a certificate authority and expose them to the CA.
+            self.ca_client.request_server_certificate(common_name, sans)
+
+        def _on_tls_config_ready(self, event):
+            # When TLS config is ready, a CA certificate, requested certificate and key will be
+            # available from an instance of CAClient. It can be written to the target files and
+            # used by the target application.
+            self.TLS_KEY_PATH.write_bytes(self.ca_client.key.private_bytes(
+                encoding=serialization.Encoding.PEM,
+                format=serialization.PrivateFormat.TraditionalOpenSSL,
+                encryption_algorithm=serialization.NoEncryption(),
+            ))
+            self.TLS_CERT_PATH.write_bytes(
+                self.ca_client.certificate.public_bytes(encoding=serialization.Encoding.PEM))
+            self.TLS_CA_CERT_PATH.write_bytes(
+                self.ca_client.ca_certificate.public_bytes(encoding=serialization.Encoding.PEM))
+            # And then reconfigure and reload your application.
+            # ...
+"""
+
+
+from ops.framework import Object, EventBase, EventSetBase, EventSource, StoredState
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.serialization import load_pem_private_key
+from cryptography.x509 import load_pem_x509_certificate
+
+import json
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class CAAvailable(EventBase):
+    pass
+
+
+class TlsConfigReady(EventBase):
+    pass
+
+
+class CAClientEvents(EventSetBase):
+    ca_available = EventSource(CAAvailable)
+    tls_config_ready = EventSource(TlsConfigReady)
+
+
+class CAClient(Object):
+
+    on = CAClientEvents()
+    stored = StoredState()
+
+    def __init__(self, charm, relation_name):
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+        self._common_name = None
+        self._sans = None
+
+        self.stored.set_default(ca_certificate=None, key=None, certificate=None)
+
+        self.framework.observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    @property
+    def is_joined(self):
+        return self.framework.model.get_relation(self._relation_name) is not None
+
+    @property
+    def is_ready(self):
+        return all(p is not None for p in (self.stored.certificate, self.stored.key,
+                                           self.stored.ca_certificate))
+
+    @property
+    def certificate(self):
+        return load_pem_x509_certificate(self.stored.certificate.encode('utf-8'),
+                                         backend=default_backend())
+
+    @property
+    def key(self):
+        return load_pem_private_key(self.stored.key.encode('utf-8'), password=None,
+                                    backend=default_backend())
```

Comment:
Once we switch to using controller-side storage, the pattern will change but I agree that as of today the key will end up wherever `/var/lib/juju/agents/unit-<app-name>-<unit-num>/charm/` is mounted which may or may not be a secure rootfs.

Based on my previous experience, charms often store secrets in the leader data (~peer app relation data) and rely on Juju to keep them secure in transit. It's not the case for storage "at rest" but it could be which would solve the problem provided controller-side storage is used.

In my view, we could provide something like `SecureStoredState` albeit it is a very general statement to call something "secure" and will likely give a false sense of security.

Reply:
We've had a few rounds of discussion in Juju about secrets. At one point it was just going to be schema on config. but the more I've thought about it, I like having something separately called Secrets. Mostly because you don't then accidentally log them/write them insecurely because they happen to be part of another data bag. Having it expressly named as a separate thing makes it clearer that they should be filtered.